Adding UIA support for WebBrowser

dkazennov opened the pull request:
Partially implements https://github.com/dotnet/winforms/issues/3421
For .Net 7

## Proposed changes

- Updated "SupportsUiaProviders" flag to WebBrowser
- Added and implemented WebBrowser.WebBrowserAccessibleObject
- Added unit tests
- Testing application 
[WebBrowserWinFormsApp.zip](https://github.com/dotnet/winforms/files/8919406/WebBrowserWinFormsApp.zip)


## Customer Impact

## Screenshots
### Before
![before](https://user-images.githubusercontent.com/102961955/174087777-ddd2fd31-17fa-48c3-8c9e-a8e9acafed37.png)

ProviderDescription:	"[pid:22832,providerId:0x402AA Main:Nested [pid:27400,providerId:0x402AA Main(parent link):Microsoft: MSAA Proxy (unmanaged:UIAutomationCore.dll)]; Hwnd(parent link):Microsoft: HWND Proxy (unmanaged:uiautomationcore.dll)]"
	
### After
![after](https://user-images.githubusercontent.com/102961955/174087809-9931a198-f0ec-45b4-8d8c-16bfaf77d464.png)

ProviderDescription:	"[pid:25792,providerId:0x70BA2 Main:Nested [pid:30696,providerId:0x70BA2 Main(parent link):Unidentified Provider (unmanaged:coreclr.dll)]; Hwnd(parent link):Microsoft: HWND Proxy (unmanaged:uiautomationcore.dll)]"

## Regression? 

- No

## Risk

- Minimal







## Test methodology

- Unit tests
- Manual with custom application

## Accessibility testing

- Inspect
- Accessibility Insights


 

## Test environment

dotnet : 7.0.0-preview.4.22229.4


###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7314)

RussKie commented:
To be honest I don't think we should be spending any time updating any of WebBrowser API. 
1. IE is [officially retired and out of support](https://blogs.windows.com/windowsexperience/2022/06/15/internet-explorer-11-has-retired-and-is-officially-out-of-support-what-you-need-to-know/)  as of  June 15, 2022.
2. We're planning to deprecated all related API (see https://github.com/dotnet/winforms/issues/6964).

/cc: @merriemcgaw @Tanya-Solyanik

dkazennov commented:
Fixed tests: ControlAccessibleObject_GetPropertyValue_AccessKey_ReturnExpected
added typeof(WebBrowser) to s_controlsIgnoringTextChangesForTests (since WebBrowserBase has public override string Text property which throws exception in setter).

dkazennov commented:
1.  WebBrowser class implements UIA provider now. New class WebBrowserAccessibleObject passes unit tests.
2. Custom application works with WebBrowser control (both original and changed) as intended.
3. New unit tests work as intended.

I believe that everything works now. So can we keep this fix as it is already complete?

vladimir-krestov commented on src/System.Windows.Forms/src/System/Windows/Forms/WebBrowser.WebBrowserAccessibleObject.cs:
```diff
@@ -0,0 +1,30 @@
+            internal override object? GetPropertyValue(UiaCore.UIA propertyID)
+                => propertyID switch
+                {
+                    UiaCore.UIA.AutomationIdPropertyId
```
Please remove, it will be implemented in https://github.com/dotnet/winforms/pull/7206

dkazennov replied:
> Please remove, it will be implemented in #7206

In #7206 it will be UiaCore.UIA.AutomationIdPropertyId => AutomationId instead. Returning Owner.Name is pre-7206.
Or should we change it pre-emptively?

dkazennov replied:
> Please remove, it will be implemented in #7206

Since #7206 is not yet implemented right now, I propose we don't change this method until #7206 is implemented (or if this issue is implemented first we will change WebBrowserAccessibleObject in #7206).

RussKie replied:
Just a friendly reminder to not lose this

dkazennov replied:
> Just a friendly reminder to not lose this

Thank you! It will be done now )

vladimir-krestov commented on src/System.Windows.Forms/src/System/Windows/Forms/WebBrowser.WebBrowserAccessibleObject.cs:
```diff
@@ -0,0 +1,30 @@
+    {
+        internal class WebBrowserAccessibleObject : ControlAccessibleObject
+        {
+            public WebBrowserAccessibleObject(WebBrowser owner) : base(owner)
+            {
+            }
```
```suggestion
            public WebBrowserAccessibleObject(WebBrowser owner) : base(owner)
            { }
```

RussKie replied:
boo :)

vladimir-krestov commented on src/System.Windows.Forms/src/System/Windows/Forms/WebBrowser.WebBrowserAccessibleObject.cs:
```diff
@@ -0,0 +1,30 @@
+                        => Owner.Name,
+                    UiaCore.UIA.HasKeyboardFocusPropertyId
+                        => Owner.Focused,
+                    UiaCore.UIA.IsKeyboardFocusablePropertyId
```
Please remove, it is implemented in the base class

dkazennov replied:
> Please remove, it is implemented in the base class

Ok!

vladimir-krestov commented on src/System.Windows.Forms/src/System/Windows/Forms/WebBrowser.cs:
```diff
@@ -64,6 +64,8 @@ public WebBrowser() : base("8856f961-340a-11d0-a96b-00c04fd705a2")
         // Public properties:
         //
 
+        internal override bool SupportsUiaProviders => true;
```
Please move it in alphabet order

vladimir-krestov commented on src/System.Windows.Forms/src/System/Windows/Forms/WebBrowser.cs:
```diff
@@ -1130,6 +1132,8 @@ protected override WebBrowserSiteBase CreateWebBrowserSiteBase()
             return new WebBrowserSite(this);
         }
 
+        protected override AccessibleObject CreateAccessibilityInstance() => new WebBrowserAccessibleObject(this);
```
ditto

vladimir-krestov reviewed (commented):
I don't see risk of implementing this fast fix, even if IE is outdated.
The main reason - https://github.com/dotnet/winforms/pull/3037. We need to provide a correct accessibility tree for implemented apps, that can use `WebBrowser`.
@merriemcgaw, @Tanya-Solyanik, your thoughts?

RussKie commented:
/azp run

RussKie reviewed (commented):
LGTM

dkazennov commented:
CTI requested

Olina-Zhang commented:
Tested this PR with no new issue found.

Philip-Wang01 commented:
Verified on latest .NET core 7.0 build: .NET 7.0.100-preview.7.22330.8, issue is fixed. Screenshot as below:
![image](https://user-images.githubusercontent.com/86937911/176846656-14986e1c-0554-4fbb-a754-7909204dee06.png)

Cassie-Li01 commented:
Verified on .NET 7.0.100-preview.7.22370.3 TP build, issue is fixed, screenshot as below:
![image](https://user-images.githubusercontent.com/96758262/180371037-22f81199-001e-46b8-865c-0d47e52424ab.png)
